spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable errors** occur when a variable reference cannot be resolved.
+
+- **Selection errors** occur when message selection fails.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+  - **Missing Fallback errors** occur when no Variant is selected
+    due to the message not including a Variant with only catch-all keys.
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+During selection, an expression handler must only emit Resolution and Selection errors.
+During formatting, an expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+
+When an error occurs in the syntax or resolution of an Expression or MarkupStart Option,
+the Expression or MarkupStart in question is processed as if the option were not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Selector error is emitted.
+When selection fails to match any Variant,
+an empty string is used as the formatted string representation of the message
+and a Missing Fallback error is emitted.
+
+When an error occurs in a Placeholder that is being formatted,
+the fallback string representation of the Placeholder
+always starts with U+007B LEFT CURLY BRACKET `{`
+and ends with U+007D RIGHT CURLY BRACKET `}`.
+Between the brackets, the following contents are used:
+
+- Expression with Literal Operand: U+0028 LEFT PARENTHESIS `(`
+  followed by the value of the Literal,
+  and then by U+0029 RIGHT PARENTHESIS `)`
+- Expression with Variable Operand: U+0024 DOLLAR SIGN `$`
+  followed by the Variable Name of the Operand
+- Expression with no Operand: U+003A COLON `:` followed by the Expression Name
+- Markup start: U+002B PLUS SIGN `+` followed by the MarkupStart Name
+- Markup end: U+002D HYPHEN-MINUS `-` followed by the MarkupEnd Name
+- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character
+
+For example, the formatted string representation of the expression `{$foo :bar}`
+would be `{$foo}` if the variable could not be resolved.
+
+The formatted string representation of a message with an unrecoverable syntax error
+is the concatenation of U+007B LEFT CURLY BRACKET `{`,
+a string identifier for the message,
+and U+007D RIGHT CURLY BRACKET `}`.
+If an identifier is not available,
+it is replaced with the U+FFFD REPLACEMENT CHARACTER `�` character,
+resulting in the string `{�}`.
```

Comment:
@eemeli I think this is better, but I would suggest that we start to hew more towards a mechanical description. Specifically, I would define the "fallback string" and then reference it. Something like:

> Let `fallbackString` be `U+FFFD REPLACEMENT CHARACTER`
> If /\* when is it otherwise defined?? \*/ let `fallbackString` be /\* that value \*/
> Let `returnValue` be the sequence `U+007B` + `fallbackString` + `U+007D`